journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournal.java

```diff
@@ -521,26 +525,27 @@ private boolean handleSegmentCorruption(
   }
 
   private void deleteDeferredFiles() {
-    final File[] files =
-        directory.listFiles(
-            file -> file.isFile() && JournalSegmentFile.isDeletedSegmentFile(name, file.getName()));
-
-    if (files == null) {
-      throw new IllegalStateException(
-          String.format(
-              "Could not list files in directory '%s'. Either the path doesn't point to a directory or an I/O error occurred.",
-              directory));
+    try (final DirectoryStream<Path> segmentsToDelete =
+        Files.newDirectoryStream(
+            directory.toPath(),
+            path -> JournalSegmentFile.isDeletedSegmentFile(name, path.getFileName().toString()))) {
+      segmentsToDelete.forEach(
```

Comment:
:wrench: If you want to avoid the double try catch, I think `DirectoryStream` is an `Iterable`, so you could just put it in a `for()` loop and let the outer catch handle the exception.

Reply:
:wrench: Another option could be to use `Files.list` - this returns a plain `Stream`. This would simplify the construction of the stream (so you can add the filter inside the try block), but then it's not an `Iterable` and you would definitely need the double try/catch :sweat_smile:

Reply:
Two try-catches are useful here as the error messages are slightly different. The inner catch has more details on which file caused the exception.

Reply:
Oops, I guess I read too quickly as I was sure they were the same :sweat_smile: